src/NUnitFramework/framework/Internal/TestNameGenerator.cs

```diff
@@ -0,0 +1,457 @@
+    /// <summary>
+    /// TestNameGenerator is able to create test names according to
+    /// a coded pattern.
+    /// </summary>
+    public class TestNameGenerator
+    {
+        private List<NameFragment> _fragments = new List<NameFragment>();
+
+        /// <summary>
+        /// Construct a TestNameGenerator
+        /// </summary>
+        /// <param name="pattern">The pattern used by this generator.</param>
+        public TestNameGenerator(string pattern)
+        {
+            // Build a list of actions so this generator can be applied to
+            // multiple types and methods.
+
+            int start = 0;
+            while (start < pattern.Length)
+            {
+                int lcurly = pattern.IndexOf('{', start);
+                if (lcurly < 0) // No more substitutions in pattern
+                    break;
+
+                int rcurly = pattern.IndexOf('}', lcurly);
+                if (rcurly < 0)
+                    break;
+
+                if (lcurly > start) // Handle fixedixed text before curly brace
+                    _fragments.Add(new FixedTextFragment(pattern.Substring(start, lcurly - start)));
+
+                string token = pattern.Substring(lcurly, rcurly - lcurly + 1);
+
+                switch (token)
+                {
+                    case "{m}":
+                        _fragments.Add(new MethodNameFragment());
+                        break;
+                    case "{n}":
+                        _fragments.Add(new NamespaceFragment());
+                        break;
+                    case "{c}":
+                        _fragments.Add(new ClassNameFragment());
+                        break;
+                    case "{C}":
+                        _fragments.Add(new ClassFullNameFragment());
+                        break;
+                    case "{M}":
+                        _fragments.Add(new MethodFullNameFragment());
+                        break;
+                    case "{a}":
+                        _fragments.Add(new ArgListFragment(40));
+                        break;
+                    case "{0}":
+                    case "{1}":
+                    case "{2}":
+                    case "{3}":
+                    case "{4}":
+                    case "{5}":
+                    case "{6}":
+                    case "{7}":
+                    case "{8}":
+                    case "{9}":
+                        int index = token[1] - '0';
+                        _fragments.Add(new ArgumentFragment(index, 40));
+                        break;
+                    default:
+                        char c = token[1];
+                        if (token.Length >= 5 && token[2] == ':' && (c == 'a' || char.IsDigit(c)))
+                        {
+                            int length;
+
+                            // NOTE: The code would be much simpler using TryParse. However,
+                            // that method doesn't exist in the Compact Framework.
```

Comment:
This is similar to our discussion on #575. I'm always torn between making the code cleaner by creating our own version of stuff like this and forcing the developer to confront the differences across platforms. @oznetmaster What do you think about this case?

We recently did exactly what I did to resolve the same issue: forced exception to be thrown. It would be good to have a "standard" approach.

We may want to table this to allow time to come to a standard approach we are willing to apply across all platforms. Throwing an exception for bad input doesn't seem all that bad. I can go either way though.